<a href="https://colab.research.google.com/github/AlienX77-cmd/Python-for-Business-and-Finance/blob/main/Part_2_Stats_Exercise_03.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Coding Exercises (Part 2)

Now, you will have the opportunity to practice what you have learned. <br>
__Follow the instructions__ and insert your code! 

The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need some help or if you want to check your code, you can also have a look at the __solutions__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Exercise 3: Sampling and Estimation

1. Import numpy and the S&P500 price returns (in %) for the year 2018 for the complete population (size = 500). 

In [1]:
# run the cell
import numpy as np
np.set_printoptions(precision=4, suppress= True)

In [2]:
# run the cell
pop = np.loadtxt("SP500_2018.csv", delimiter = ",", usecols = 1)
pop

array([ -1.5081, -38.8238,  55.9434,  -7.6819,  -5.7802, -19.4511,
        69.7722,  24.5663,  -8.6681,  27.3282,  -4.0099,   1.7465,
        10.2739, -40.8947,  21.3679,  10.0865,   1.5767,  31.7636,
         2.4151, -18.8165, -35.1796,  -0.5262, -12.5744,  14.6018,
        -6.1347, -39.9953,  -6.2559, -17.9567, -21.6694,  -1.2695,
       -18.7139, -36.6588, -21.1864,  73.3463,  -7.9205, -53.2229,
         9.678 , -39.1338,  11.3268,  26.3837, -12.6581,  -3.1167,
        15.1415,   7.2388, -30.3198, -37.6125,  -2.06  ,  -8.3485,
       -27.6435, -12.7498, -38.4954,   6.7179, -26.0897,  -3.2173,
        -1.2884, -22.6624,  -1.6067,  -4.924 ,  17.9428,   7.2802,
       -17.378 ,   0.8818, -13.6163, -25.0183,   3.7046, -32.6794,
       -13.6877, -32.2377,  -7.8633, -13.3866,  -1.2608, -24.4262,
        18.362 , -16.8734,   5.0232,   1.9473,  39.8548, -32.4525,
       -13.9199, -30.3454, -43.7749, -27.2401, -20.2881, -12.5368,
       -22.7787,  -7.8966, -26.4237,  -2.6034, -26.6536,   3.6

2. Draw a __random sample__ with sample size __50__ and __save__ the sample. Use the random __seed 123__.

In [3]:
sample_size = 50

In [4]:
np.random.seed(123)
sample = np.random.choice(pop, sample_size, replace = False)

In [5]:
sample

array([ -3.189 ,  -0.7402, -10.0045,  -0.6494,  -6.7991, -10.3929,
       -19.4511, -21.3489, -23.8865, -24.194 ,  19.9705,  -2.8071,
        -7.9205,   2.8771, -42.3216,   1.7465, -54.1992, -20.5551,
        -0.0258,  36.4959, -22.8224, -57.5208, -10.6338,  -9.1523,
         9.707 , -21.1643, -23.9187,   7.2802, -45.8564,  33.4487,
         0.0822,  -7.8966, -27.5657,   9.678 , -15.3846,  35.2119,
         7.1421,  17.2502,  73.3463,  35.6039,  18.6085,  27.3282,
        -0.8907,   4.4209, -27.6435, -18.8478,  10.0865,  22.4127,
        -6.1347, -20.2881])

++++ __From this point, assume that you only have the sample. You don´t know anything about the population.__++++ 

3. Calculate and save the __sample mean__ (point estimate of the mean).

In [6]:
point_est_mean = sample.mean()
point_est_mean

-3.8301530353346083

(result: -3.8301530353346083)

4. Calculate and save the __sample standard deviation__ (point estimate of the std).

In [7]:
point_est_std = sample.std(ddof = 1)
point_est_std

24.56795822088298

(result: 24.56795822088298)

5. Calculate the __Standard Error__ of the sample mean.

In [8]:
standard_error = point_est_std / np.sqrt(sample_size)
standard_error

3.4744339715788284

(result: 3.4744339715788284)

6. Estimate the (__equally-weighted__) mean return for the whole S&P 500 population for the year 2018 by constructing a __90% Confidence Interval__.

In [9]:
import scipy.stats as stats

In [12]:
conf = 0.9


In [13]:
stats.t.interval(conf, loc = point_est_mean, scale = standard_error, df = sample_size - 1)

(-9.655218409282309, 1.9949123386130907)

(result: (-9.655218409282309, 1.9949123386130907))

7. __Same as Q6__. This time, construct the 90% Confidence Interval with __Bootstrapping__. Use __100,000 simulations__ and the random __seed 123__.

In [14]:
simulations = 100000

In [15]:
np.random.seed(123)
bootstrap = []
for i in range(simulations):
	bootstrap.append(np.random.choice(sample, size = sample_size, replace = True).mean())

In [16]:
bootstrap

[-3.935833651990379,
 -6.496999972968056,
 3.2007691286476017,
 -5.122473441877855,
 -8.36618662340559,
 -5.944702809771454,
 1.9443665699316472,
 -4.332207600190588,
 -10.17418741818757,
 1.0767255631400237,
 -0.24184546142679728,
 -6.233485879840373,
 -9.10695604666338,
 5.134460138979331,
 -4.752975174596158,
 -2.0762345589150097,
 -2.401156990057205,
 -7.91018553629927,
 0.4783555245366553,
 -2.322181115291327,
 -4.708805006564257,
 -5.880065065954564,
 -0.26379065228606824,
 -1.5463157904873719,
 -2.2704084507535915,
 -3.4620877126281875,
 -7.030830774695699,
 -7.698407934492674,
 -5.320737472096145,
 -6.015611622938131,
 -9.958485068556037,
 -0.9566018003513463,
 1.105699865369129,
 -3.331663929032915,
 -0.4769571022913699,
 -5.032851290397858,
 1.161556569248159,
 -6.395769476563086,
 -6.0713099367075385,
 -7.164353245890052,
 -4.009342020589453,
 -2.280877318900457,
 -2.0716340211996247,
 -0.1926900415268155,
 -6.351288708760682,
 3.5643137549687465,
 -2.966956277463583,
 -3.09

In [17]:
np.percentile(bootstrap, [5, 95])

array([-9.404 ,  1.9449])

(result: array([-9.404 ,  1.9449]))

# Well Done!

---------------------------------------------------------------------------------------------------------------------

# Solutions (Stop here if you want to code on your own!)

# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

1. Import numpy and the S&P500 price returns (in %) for the year 2018 for the complete population (size = 500). 

In [ ]:
# run the cell
import numpy as np
np.set_printoptions(precision=4, suppress= True)

In [ ]:
# run the cell
pop = np.loadtxt("SP500_2018.csv", delimiter = ",", usecols = 1)
pop

2. Draw a __random sample__ with sample size __50__ and __save__ the sample. Use the random __seed 123__.

In [ ]:
sample_size = 50

In [ ]:
np.random.seed(123)
sample = np.random.choice(pop, sample_size, replace = False)

In [ ]:
sample

++++ __From this point, assume that you only have the sample. You don´t know anything about the population.__++++ 

3. Calculate and save the __sample mean__ (point estimate of the mean).

In [ ]:
point_est_mean = sample.mean()
point_est_mean

(result: -3.8301530353346083)

4. Calculate and save the __sample standard deviation__ (point estimate of the std).

In [ ]:
point_est_std = sample.std(ddof = 1)
point_est_std

(result: 24.56795822088298)

5. Calculate the __Standard Error__ of the sample mean.

In [ ]:
standard_error = point_est_std / np.sqrt(sample_size)
standard_error

(result: 3.4744339715788284)

6. Estimate the (__equally-weighted__) mean return for the whole S&P 500 population for the year 2018 by constructing a __90% Confidence Interval__.

In [ ]:
import scipy.stats as stats

In [ ]:
conf = 0.9

In [ ]:
stats.t.interval(conf, loc = point_est_mean, scale = standard_error, df = sample_size - 1)

(result: (-9.655218409282309, 1.9949123386130907))

7. __Same as Q6__. This time, construct the 90% Confidence Interval with __Bootstrapping__. Use __100,000 simulations__ and the random __seed 123__.

In [ ]:
sims = 100000

In [ ]:
np.random.seed(123)
bootstrap = []
for i in range(sims):
    bootstrap.append(np.random.choice(sample, size = sample_size, replace = True).mean())

In [ ]:
bootstrap

In [ ]:
np.percentile(bootstrap, [5, 95])

(result: array([-9.404 ,  1.9449]))